contracts/abstract/AbstractNFTRewardDelegate.sol

```diff
@@ -0,0 +1,391 @@
+
+  //*********************************************************************//
+  // ---------------------- ITokenSupplyDetails ------------------------ //
+  //*********************************************************************//
+
+  function totalSupply() public view override returns (uint256) {
+    return _supply;
+  }
+
+  function tokenSupply(uint256 _tokenId) public view override returns (uint256) {
+    return _ownerOf[_tokenId] != address(0) ? 1 : 0;
+  }
+
+  function totalOwnerBalance(address _account) public view override returns (uint256) {
+    if (_account == address(0)) {
+      revert INVALID_ADDRESS();
+    }
+
+    return _balanceOf[_account];
+  }
+
+  function ownerTokenBalance(address _account, uint256 _tokenId)
+    public
+    view
+    override
+    returns (uint256)
+  {
+    return _ownerOf[_tokenId] == _account ? 1 : 0;
+  }
+
+  //*********************************************************************//
+  // ----------------------------- ERC721 ------------------------------ //
+  //*********************************************************************//
+
+  /**
+    @notice Returns the full URI for the asset.
+  */
+  function tokenURI(uint256 tokenId) public view override returns (string memory) {
+    if (_ownerOf[tokenId] == address(0)) {
+      revert INVALID_TOKEN();
+    }
+
+    if (address(_tokenUriResolver) != address(0)) {
+      return _tokenUriResolver.tokenURI(tokenId);
+    }
+
+    return bytes(_baseUri).length > 0 ? string(abi.encodePacked(_baseUri, tokenId.toString())) : '';
+  }
+
+  /**
+    @notice Returns the contract metadata uri.
+  */
+  function contractURI() public view override returns (string memory contractUri) {
+    contractUri = _contractUri;
+  }
+
+  /**
+    @notice
+    Transfer tokens to an account.
+
+    @param _to The destination address.
+    @param _id NFT id to transfer.
+  */
+  function transfer(address _to, uint256 _id) public override {
+    if (!transferrable) {
+      revert NON_TRANSFERRABLE();
+    }
+    transferFrom(msg.sender, _to, _id);
+  }
+
+  /**
+    @notice
+    Transfer tokens between accounts.
+
+    @param _from The originating address.
+    @param _to The destination address.
+    @param _id The amount of the transfer, as a fixed point number with 18 decimals.
+  */
+  function transferFrom(
+    address _from,
+    address _to,
+    uint256 _id
+  ) public override {
+    if (!transferrable) {
+      revert NON_TRANSFERRABLE();
+    }
+    super.transferFrom(_from, _to, _id);
+  }
+
+  /**
+    @notice
+    Confirms that the given address owns the provided token.
+   */
+  function isOwner(address _account, uint256 _id) public view override returns (bool) {
+    return _ownerOf[_id] == _account;
+  }
+
+  // TODO: this will cause issues for some price resolvers
+  function mint(address _account) external override onlyOwner returns (uint256 tokenId) {
+    if (_supply == _maxSupply) {
+      revert SUPPLY_EXHAUSTED();
+    }
+
+    tokenId = _supply;
```

Comment:
do this assignment before the _supply==_maxSupply and then use tokenId to compare (that's another double sread currently)